EIPS/eip-5409.md

```diff
@@ -0,0 +1,65 @@
+description: Allow EIP-1155 to represent Non-Fungible Tokens (tokens who have a unique owner)
+author: Ronan Sandford (@wighawag)
+discussions-to: https://ethereum-magicians.org/t/eip-5409-non-fungible-token-extension-for-eip-1155/10240
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-07-23
+requires: 165, 1155
+---
+
+## Abstract
+
+This standard is an extension of [EIP-1155](./eip-1155.md). It proposes an additional function, `ownerOf` to allow EIP-1155 to support Non-Fungible Tokens (tokens that have a unique owner). By implementing this extra function, EIP-1155 can benefit from [EIP-721](./eip-721.md) core functionality without implementing the full (and less efficient) EIP-721 spec in the same contract. In particular, this extension allows EIP-1155 tokens of supply one to have an owner and have it exposed.
+
+## Motivation
+
+Currently, EIP-1155 does not allow an external caller to detect whether a token is truly unique (can have only one owner) or fungible. This is because EIP-1155 do not expose a mechanism to detect whether a token will have its supply remain to be "1". Furthermore, it does not let an external caller retrieve the owner directly on-chain.
+
+The EIP-1155 spec does mention the use of split id to represent non-fungible tokens, but this requires a pre-established convention that is not part of the standard. Plus it would not be as fully featured as the `ownerOf` function also present in the EIP-721 standard.
+
+The ability to get the owner of a token offer great benefit, including the ability for the owner to associate data to it.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### Contract Interface
+
+```solidity
+interface IERC1155OwnerOf {
+
+    /// @notice Find the owner of an NFT
+    /// @dev The zero address indicates that there is no owner: either the token does not exist or it is not an NFT (supply potentially bigger than 1)
+    /// @param tokenId The identifier for an NFT
+    /// @return The address of the owner of the NFT
+    function ownerOf(uint256 tokenId) external view returns (address);
+}
+```
+
+The `ownerOf(uint256 tokenId)` function MAY be implemented as `pure` or `view`.
+
+The `supportsInterface` method MUST return `true` when called with `0x6352211e`.
+
+## Rationale
+
+Having the ability to fetch the owner for EIP-1155 tokens that have one, allows us to use them in most of the EIP-721 ecosystem, without the need to implement the whole EIP-721 spec which comes with inefficiencies.
+
+`ownerOf` does not throw when a token does not exist (or does not have an owner). This simplifies the handling of such a case. Since it would be a security risk to assume all EIP-721 implementation would throw, it should not break compatibility with contract handling EIP-721 when dealing with this EIP-1155 extension.
```

Comment:
This is incorrect. EIP-721 specifically states that `ownerOf` a `tokenId` assigned to the zero address or a nonexistent NFT throws.

Reply:
Here I describe the behavior of `ownerOf` in EIP-5409, not the behavior in EIP-721

Reply:
> Since it would be a security risk to assume all EIP-721 implementation would throw

It is not a security risk, as all EIP-721 implementations MUST throw.

I therefore also propose that `ownerOf` throw too. Small things like these have resulted in major annoyances down the line.

Reply:
My point is that a smart contract that intends to support ERC721 generically would take a risk when they assume all ERC721 implementation throw when `ownerOf` is called with a non-existent token (There is a chance some invalid implementation exists maliciously or not). 

One reason for not throwing is that it makes checking for existence cumbersome as you have to catch the error.
But to be honest, I am not too attached to not throwing if that helps. Do you have links to examples of the "small things" you mentioned?

Note that If we go we decide to throw, it must throw in both the following situations:
1. the token is a fungible token (it has no owner) 
2. or the token does not exist

Reply:
Having to catch it is a *good* thing. Exceptions are *good.* It's when there's not obviously a problem and your program misbehaves in a subtle but extremely problematic way that major issues arise.

Reply:
Well i never said throw is bad as a general statement. It all depends on the context.

In this particular one, i gave my reasoning